src/token-service/metadata-token-service.ts

```diff
@@ -11,68 +11,102 @@ const DEFAULT_OPTIONS: Options = {
     },
 };
 
+import {
+    TOKEN_MINIMUM_LIFETIME_MARGIN_MS,
+    TOKEN_LIFETIME_LEFT_TO_REFRESH_PCT,
+    MAX_ATTEMPTS_NUMBER_TO_GET_TOKEN_IN_INITIALIZE
+} from './metadata-token-service.consts'
 export class MetadataTokenService implements TokenService {
     private readonly url: string;
     private readonly opts: Options;
     private token?: string;
+    private tokenExpired = 0;
+    private tokenTimeToRefresh = 0;
```

Comment:
There are two key time points for updating a token - when 10% of the lifetime has passed and when there are 15 minutes of lifetime left.  To avoid repeating these calculations in different places (which can be a source of errors), it is easier to calculate them once when receiving a token